src/spike/constructSignatureString.js

```diff
@@ -0,0 +1,47 @@
+const {
+  HEADERS_FIELD_PREFIX,
+  SIGNATURE_FIELD_PREFIX,
+  KEY_ID_FIELD_PREFIX,
+  ALGORITHM_FIELD_PREFIX,
+  DELIMITER,
+  NEWLINE,
+  SPACE,
+  COMMA
+} = require('./constants')
+
+function joinField (key, value) {
+  return `${key}"${value}"`
+}
+
+function headerFieldReducer (accumulator, [key]) {
+  return accumulator.concat(key, SPACE)
+};
+
+function signatureHeaderInputReducer (accumulator, [key, value]) {
+  return accumulator.concat(key, DELIMITER, value, NEWLINE)
+};
+
+module.exports = function constructSignatureString (request, opts) {
+  const {
+    secret,
+    constructSignature,
+    keyId,
+    algorithm,
+    signingAlgorithm,
+    fields
+  } = opts
+
+  const signedHeaders = fields.reduce(signatureHeaderInputReducer, '').trim()
+  const signature = constructSignature(secret, signingAlgorithm, signedHeaders)
+  const headerField = joinField(HEADERS_FIELD_PREFIX, fields.reduce(headerFieldReducer, '').trim())
+  const keyIdField = joinField(KEY_ID_FIELD_PREFIX, keyId)
+  const algorithmField = joinField(ALGORITHM_FIELD_PREFIX, algorithm)
+  const signatureField = joinField(SIGNATURE_FIELD_PREFIX, signature)
+
+  return [
+    keyIdField,
+    algorithmField,
+    headerField,
+    signatureField
+  ].join(COMMA).trim()
+}
```

Comment:
The entire signature string also needs to contain the values of `created` and `expires` if they were used in the signature. 

e.g.           

`Signature: 'keyId="test-key-a", algorithm="hs2019", headers="(request-target) (created) (expires) host digest content-type", signature="HFIn7RU3zvfFHeyOjsRlMzUQ18prW+KE61ikKKKoyAlcxGUdlBF/sruA+VznOhQNlWh3J4y3tZc8aDa0TxRBEg==", created=1602872645, expires=1602872945'`

This should probably be conditional based on if they are present in the `fields` array.

Reply:
yep, that is on the roadmap (in my head 😬 )